Use block icons in media placeholders

ZebulanStanphill opened the pull request:
## Description
Fixes #9642.

This PR changes the media placeholders of the media blocks to use the corresponding icon of each block, instead of using Dashicons. Since the icons are now being used by both the `index.js` and (if it exists) `edit.js` of each of these blocks, I have moved these icons into `icon.js` files for each block, so the icons are not defined twice and can be imported by multiple other modules.

List of blocks this PR affects:
- Audio
- Cover
- File
- Gallery
- Image
- Media & Text
- Video

I also took the opportunity to reorganize the imports in the files I modified: they are now organized alphabetically, and imports that were previously incorrectly listed under "Internal dependencies" are now listed under "WordPress dependencies".

## Screenshot
![image](https://user-images.githubusercontent.com/19592990/48384717-35bd5300-e6b1-11e8-8044-b65c347f3a82.png)

## Related issues and PRs
- Replaces #9646

ZebulanStanphill committed:
Use block icons in media placeholders

ZebulanStanphill commented:
Tests are passing now. Thanks for restarting them, @earnjam! :slightly_smiling_face:

earnjam commented:
They feel just a bit crowded to the text compared to the old icons. Flagged for design feedback

See screenshots for comparison:

<img width="300" alt="screen shot 2018-11-13 at 11 29 50 am" src="https://user-images.githubusercontent.com/4267290/48427795-94c5ab00-e737-11e8-80ba-ad3ee2c3acca.png"> <img width="300" alt="screen shot 2018-11-13 at 11 29 09 am" src="https://user-images.githubusercontent.com/4267290/48427793-92fbe780-e737-11e8-9035-87a78ac43b93.png">

melchoyce commented:
Good catch, @earnjam — the icons definitely need some additional space. Does Gutenberg not automatically provide the correct spacing between icon + label for placeholder blocks? I assumed it would take care of that kind of thing for folks.

ZebulanStanphill commented:
@melchoyce I had assumed it would as well. I am using the `<BlockIcon />` component. If I don't use it and just use the `<SVG>` component directly, then the icons grow ridiculously big. Should I be using a component other than the `<BlockIcon />` component?

mtias commented:
> I assumed it would take care of that kind of thing for folks.

Should be applied at the placeholder component level. Right now it addresses svgs that are dashicons, not any svg. Should be easy, I can look into it tomorrow.

mtias commented:
Fix for it here: https://github.com/WordPress/gutenberg/pull/11947

ZebulanStanphill commented:
Rebased and resolved merge conflicts.

Thanks for the margin fix PR, @mtias! :smile:

ZebulanStanphill commented:
Rebased and resolved merge conflicts again.

gziolo reviewed (commented):
Code wise it looks quite good. I must admit that I really like the fact that icons were moved to their own file. I can perform the more in-depth review as soon as this PR gets ✅ from one of the designers.

kjellr reviewed (changes requested):
Thanks for sticking with this, @ZebulanStanphill. It's nice to get these icons all synced up.  

Design-wise, this looks pretty good. Just a minor note: All of these icons seem just a little bit too low compared to the text. The icons need to be moved up 1px in order to be centered visually: 

**Current:** 
<img width="635" alt="screen shot 2019-02-04 at 9 04 59 am" src="https://user-images.githubusercontent.com/1202812/52213109-2a1a9c00-285c-11e9-9174-20e215e39b43.png">

**After:** 
<img width="670" alt="screen shot 2019-02-04 at 9 05 55 am" src="https://user-images.githubusercontent.com/1202812/52213111-2c7cf600-285c-11e9-8422-d9a607524ad2.png">


I have one other question for everyone: should we actually be using an image icon within the Media + Text block here?

<img width="683" alt="screen shot 2019-02-04 at 9 07 33 am" src="https://user-images.githubusercontent.com/1202812/52213171-53d3c300-285c-11e9-8256-ee8ce007c89b.png">

In this case, that placeholder area is for "Media", not specifically to "Media + Text" like the icon implies. 

Thanks again!

melchoyce commented:
> In this case, that placeholder area is for "Media", not specifically to "Media + Text" like the icon implies.

The "Content" on the right is the text placeholder, right?

melchoyce commented:
Oh, I misunderstood — the label and icon should _just_ be changed within the image block placeholder, got it. I think that makes sense 👍

ZebulanStanphill commented:
Rebased PR.

@kjellr Regarding the position of the icons, I checked, and it looks like it is actually the text that is not centered. Additionally, the icons themselves appear to be drawn slightly higher in their SVG viewbox than you might expect:
![untitled](https://user-images.githubusercontent.com/19592990/52253014-093d5f80-28cb-11e9-9f18-d19ef7ba02a2.png)
![image](https://user-images.githubusercontent.com/19592990/52253178-ed868900-28cb-11e9-9cd9-a9398fece3ed.png)

I don't know what the best way to adjust this would be. I don't think it really matters that much anyway, though.

As for the Media & Text block placeholder, if I'm not supposed to use the block icon, then what should I use? The media area accepts both images and videos.

gziolo commented on packages/block-library/src/audio/edit.js:
```diff
@@ -115,7 +121,7 @@ class AudioEdit extends Component {
 			return (
 				<MediaPlaceholder
-					icon="media-audio"
+					icon=<BlockIcon icon={ icon } />
```
Can it follow other definitions of props?:
```suggestion
					icon={ <BlockIcon icon={ icon } /> }
```

gziolo commented on packages/block-library/src/file/edit.js:
```diff
@@ -136,7 +142,7 @@ class FileEdit extends Component {
 			return (
 				<MediaPlaceholder
-					icon="media-default"
+					icon=<BlockIcon icon={ icon } />
```
Can it follow other definitions of props?:
```suggestion
					icon={ <BlockIcon icon={ icon } /> }
```

gziolo commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -220,7 +222,7 @@ class GalleryEdit extends Component {
 					{ controls }
 					<MediaPlaceholder
-						icon="format-gallery"
+						icon=<BlockIcon icon={ icon } />
```
Can it follow other definitions of props?:
```suggestion
						icon={ <BlockIcon icon={ icon } /> }
```

gziolo commented on packages/block-library/src/image/edit.js:
```diff
@@ -423,7 +425,7 @@ class ImageEdit extends Component {
 					{ controls }
 					<MediaPlaceholder
-						icon="format-image"
+						icon=<BlockIcon icon={ icon } />
```
Can it follow other definitions of props?:
```suggestion
						icon={ <BlockIcon icon={ icon } /> }
```

gziolo commented on packages/block-library/src/media-text/media-container.js:
```diff
@@ -67,7 +73,7 @@ class MediaContainer extends Component {
 		return (
 			<MediaPlaceholder
-				icon="format-image"
+				icon=<BlockIcon icon={ icon } />
```
Can it follow other definitions of props?:
```suggestion
				icon={ <BlockIcon icon={ icon } /> }
```

gziolo commented on packages/block-library/src/video/edit.js:
```diff
@@ -150,7 +152,7 @@ class VideoEdit extends Component {
 			return (
 				<MediaPlaceholder
-					icon="media-video"
+					icon=<BlockIcon icon={ icon } />
```
Can it follow other definitions of props?:
```suggestion
					icon={ <BlockIcon icon={ icon } /> }
```

gziolo reviewed (commented):
I shared suggestions for some stylistic changes. Otherwise, it looks good code wise.

_Aside:_ in retrospect, I think that we should never allow elements to be defined as an icon for the block. Instead, we should be using components:
```js
const Icon = () => <SVG ...>...</SVG>;
```
However, it's too late to change that.

kjellr commented:
> Regarding the position of the icons, I checked, and it looks like it is actually the text that is not centered [...] I don't know what the best way to adjust this would be.

I think the simplest way to fix this would be to add `align-items: center;` to the `components-placeholder__label`. That seems to properly align things for me.

> As for the Media & Text block placeholder, if I'm not supposed to use the block icon, then what should I use? The media area accepts both images and videos.

This is a tough one — I'm not sure we actually have a Material-based icon for that yet. Maybe something like this? 

<img width="281" alt="screen shot 2019-02-05 at 9 20 09 am" src="https://user-images.githubusercontent.com/1202812/52279336-594a0f80-2927-11e9-91dc-438ce019fc08.png">

([SVG](https://cloudup.com/cA67x4EL0it))

^ That could use some extra eyes though, as it's not a straight-up Material icon. I took [the `image_search` icon](https://material.io/tools/icons/?search=image_search&icon=image_search&style=baseline), and added the play icon instead of a search icon.

gziolo committed:
Add braces around BlockIcon components when used as attribute values.

Co-Authored-By: ZebulanStanphill <zebulanstanphill@gmail.com>

ZebulanStanphill committed:
Vertically center content of placeholder label.

ZebulanStanphill commented:
Addressed feedback.

@kjellr I don't know... that icon mockup still seems too image-oriented. Come to think of it... what if the media area in the Media & Text block just didn't have an icon at all? Would that be alright?

kjellr commented:
Eliminating an icon altogether seems a little weird to me. Interested in other people's takes though. 

Another possible Images/Video icon direction: 

<img width="189" alt="screen shot 2019-02-05 at 11 12 58 am" src="https://user-images.githubusercontent.com/1202812/52286865-06785400-2937-11e9-8bc5-e5a83bd79152.png">

<img width="190" alt="screen shot 2019-02-05 at 11 13 56 am" src="https://user-images.githubusercontent.com/1202812/52286917-2740a980-2937-11e9-9715-a46dc03c4688.png">

ZebulanStanphill commented:
@kjellr Ooh, I like those mockups. I think that kind of icon would work.

melchoyce commented:
I think that top icon could work — and I foresee a mixed image/video icon being useful in the future.

ZebulanStanphill committed:
Change Media & Text block placeholder icon

ZebulanStanphill commented:
@kjellr I have now updated the Media & Text block placeholder to use that new icon. :slightly_smiling_face:

gziolo reviewed (approved):
On the code level, all those changes look good. I like that icons were moved to their own file 👍 

I'm leaving final ✅ to @melchoyce and @kjellr who have better idea how those icons should integrate :)

kjellr reviewed (approved):
Looks great, thank you @ZebulanStanphill! I still may iterate on that image/video icon a little bit, but that can be handled separately. 👍 

![screen-shot-2019-02-06-at-9 58 23-am](https://user-images.githubusercontent.com/1202812/52350213-055a2c00-29f6-11e9-8319-5788cc7c070e.gif)
